fix(image-album): make sure to delete all images part of an album

jrainville opened the pull request:
Needed for https://github.com/status-im/status-desktop/issues/10171

When we delete a message, if it is an Image message, we make sure to delete all the messages part of the album.

Same thing when we get the signal to delete a message, we remove all of the messages in the album.

I might need do a check if we receive later an image from the album and it was already deleted, I need to check if that's handled.
I'll also add a test

jrainville committed:
fix(image-album): make sure to delete all images part of an album

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://status.im/docs/))?
- [ ] Have you tested changes with mobile?
- [ ] Have you tested changes with desktop?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (20)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 3815d440 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3380/1/) | 2023-04-05 15:58:38 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230405-155716-3815d4-pr3380.zip) |
| :heavy_multiplication_x: | 3815d440 | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3380/1/) | 2023-04-05 15:59:23 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3380/1//consoleText) |
| :heavy_check_mark: | 3815d440 | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3380/1/) | 2023-04-05 16:01:36 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230405-155716-3815d4-pr3380.zip) |
| :heavy_check_mark: | 3815d440 | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3380/1/) | 2023-04-05 16:01:49 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230405-155716-3815d4-pr3380.aar) |
| | | | | | | |
| :heavy_check_mark: | ed91befe | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3380/2/) | 2023-04-05 15:59:36 | ~53 sec | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230405-155842-ed91be-pr3380.zip) |
| :heavy_multiplication_x: | ed91befe | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3380/2/) | 2023-04-05 16:00:12 | ~43 sec | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3380/2//consoleText) |
| :heavy_check_mark: | ed91befe | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3380/2/) | 2023-04-05 16:04:22 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230405-160139-ed91be-pr3380.zip) |
| :heavy_check_mark: | ed91befe | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3380/2/) | 2023-04-05 16:05:07 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230405-160153-ed91be-pr3380.aar) |
| | | | | | | |
| :heavy_check_mark: | 2aaf79c9 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3380/3/) | 2023-04-05 18:40:06 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230405-183848-2aaf79-pr3380.zip) |
| :heavy_multiplication_x: | 2aaf79c9 | [#3](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3380/3/) | 2023-04-05 18:40:54 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3380/3//consoleText) |
| :heavy_check_mark: | 2aaf79c9 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3380/3/) | 2023-04-05 18:42:07 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230405-183843-2aaf79-pr3380.aar) |
| :heavy_check_mark: | 2aaf79c9 | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3380/3/) | 2023-04-05 18:43:06 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230405-183848-2aaf79-pr3380.zip) |
| | | | | | | |
| :heavy_multiplication_x: | fec35bb6 | [#4](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3380/4/) | 2023-04-06 16:42:46 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3380/4//consoleText) |
| :heavy_check_mark: | fec35bb6 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3380/4/) | 2023-04-06 16:43:02 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230406-164026-fec35b-pr3380.zip) |
| :heavy_check_mark: | fec35bb6 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3380/4/) | 2023-04-06 16:43:12 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230406-164025-fec35b-pr3380.zip) |
| :heavy_check_mark: | fec35bb6 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3380/4/) | 2023-04-06 16:44:21 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230406-164024-fec35b-pr3380.aar) |
| | | | | | | |
| :heavy_check_mark: | de0df2fd | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3380/5/) | 2023-04-06 18:00:48 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230406-175832-de0df2-pr3380.zip) |
| :heavy_multiplication_x: | de0df2fd | [#5](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3380/5/) | 2023-04-06 18:00:52 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3380/5//consoleText) |
| :heavy_check_mark: | de0df2fd | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3380/5/) | 2023-04-06 18:01:47 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230406-175832-de0df2-pr3380.zip) |
| :heavy_check_mark: | de0df2fd | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3380/5/) | 2023-04-06 18:02:52 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230406-175828-de0df2-pr3380.aar) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_multiplication_x: | ac77846d | [#6](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3380/6/) | 2023-04-13 19:29:25 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3380/6//consoleText) |
| :heavy_check_mark: | ac77846d | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3380/6/) | 2023-04-13 19:30:42 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230413-192704-ac7784-pr3380.zip) |
| :heavy_check_mark: | ac77846d | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3380/6/) | 2023-04-13 19:31:32 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230413-192704-ac7784-pr3380.zip) |
| :heavy_check_mark: | ac77846d | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3380/6/) | 2023-04-13 19:32:10 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230413-192704-ac7784-pr3380.aar) |
| | | | | | | |
| :heavy_check_mark: | e85afab9 | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3380/7/) | 2023-04-14 14:22:52 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230414-142010-e85afa-pr3380.zip) |
| :heavy_check_mark: | e85afab9 | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3380/7/) | 2023-04-14 14:23:08 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230414-142010-e85afa-pr3380.zip) |
| :heavy_check_mark: | e85afab9 | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3380/7/) | 2023-04-14 14:24:30 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230414-142010-e85afa-pr3380.aar) |
| :heavy_check_mark: | e85afab9 | [#7](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3380/7/) | 2023-04-14 14:32:03 | ~11 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3380/7//consoleText) |

borismelnik reviewed (approved):
LGTM :) but will be great to add some test for it

ilmotta commented on protocol/messenger_handler.go:
```diff
@@ -1574,21 +1574,44 @@ func (m *Messenger) HandleDeleteMessage(state *ReceivedMessageState, deleteMessa
 	// Update message and return it
-	originalMessage.Deleted = true
-	originalMessage.DeletedBy = deleteMessage.DeleteMessage.DeletedBy
+	for _, messageToDelete := range messagesToDelete {
```
@jrainville, do you know if the action of deleting messages is happening inside a transaction? If not, then although rare, it would be better if we could save all messages in one go (query) or do one by one and rollback in case one of them fail before the loop finishes.

jrainville replied:
I don't think it happens in a transaction.
As for rolling back, I'm not sure if it's necessary. Since the goal is to delete, if one of the messages fails, it's ok if the some of them are deleted IMO.
What would happen on the front is that one of the images would disappear, the user could try to delete again for the remaining ones.

ilmotta replied:
Thanks for taking a look @jrainville. It would be a bit better if we could be more consistent by default, but as you said and I agree, in the rare case that something happens the user could just try again, so it's fine.

ilmotta commented on protocol/messenger_messages.go:
```diff
@@ -157,6 +157,21 @@ func (m *Messenger) DeleteMessageAndSend(ctx context.Context, messageID string)
 		return nil, ErrInvalidDeleteTypeAuthor
 	}
 
+	var messagesToDelete []*common.Message
```
This piece of code that builds up the messages to be deleted is identical or very similar to what's in `messenger_handler.go` (I haven't checked thoroughly). I think we should at least extract it into a common function.

ilmotta commented on protocol/messenger_messages.go:
```diff
@@ -184,11 +199,16 @@ func (m *Messenger) DeleteMessageAndSend(ctx context.Context, messageID string)
-	if err != nil {
-		return nil, err
+	response := &MessengerResponse{}
+	for _, messageToDelete := range messagesToDelete {
```
As @borismelnik mentioned already, I think we should cover the new code with a bit of testing in this PR, or in a follow-up PR if this one needs to be merged asap.

Edit: Sorry, I didn't see you mentioned in the PR description about adding tests.

jrainville commented:
Thanks for the reviews. As stated in the description, I'm working on adding tests right now. I'll push them in another commit this afternoon.

jrainville commented:
I added a test for deleting an album. I'll add another one later in case not all messages arrived when we got a deletion.
I also need to address @ilmotta 's comment.
I'll do it tomorrow, there is no rush to merge this. I need to tackle an issue for our release in the meantime

jrainville committed:
test(delete): add test where the signal to delete images is after

also adds the handling of deleteForMe

jrainville committed:
fix(delete): add album deletion handling for deleteForMe

jrainville commented:
Alright, I addressed the comments and implemented the tests.
We don't support DeleteForMe on Desktop yet, so I wasn't able to test that code in the client itself, so if someone from mobile could check that I didn't break it (it's pretty much the same as Delete, so it should work), it would be great. Thanks.

ilmotta commented:
> Alright, I addressed the comments and implemented the tests. We don't support DeleteForMe on Desktop yet, so I wasn't able to test that code in the client itself, so if someone from mobile could check that I didn't break it (it's pretty much the same as Delete, so it should work), it would be great. Thanks.

@jrainville, I can test this next Monday. Today is a bit tight because tomorrow I'm off. Would that be okay?

ilmotta commented on protocol/messenger_test.go:
```diff
@@ -2319,6 +2321,87 @@ func (s *MessengerSuite) TestResendExpiredEmojis() {
+}
+
+/*
+func testSendImages(s *suite.Suite, messageCount int, ourChat *Chat, theirMessenger *Chat) []*common.Message {
```
Do we really need to keep this commented test?

jrainville commented:
> I can test this next Monday. Today is a bit tight because tomorrow I'm off. Would that be okay?

yeah that's good. I won't pull this into our release in the end. It's not a crucial fix and it touches more code than I thought it would.

jrainville committed:
test(delete): add test that deletes a message part of an album


opps

ilmotta commented:
@jrainville I just tried to test this branch, but I'm getting an error while using the chat because the endpoint `ChatMentionCheckMentions` is not available. Could you rebase so I can proceed with the tests in the mobile app?

jrainville commented:
@ilmotta done

ilmotta commented:
> @ilmotta done

@jrainville Tested on mobile and it behaves as in `develop`. Double-checked Delete for me, and all images are deleted from the album :rocket:

jrainville committed:
linting
